ufl/algorithms/remove_complex_nodes.py

```diff
@@ -0,0 +1,36 @@
+# -*- coding: utf-8 -*-
+"""Algorithm for removing conj, real, and imag nodes
+from a form for when the user is in 'real mode'"""
+
+from ufl.corealg.multifunction import MultiFunction
+from ufl.constantvalue import ComplexValue, FloatValue
+from ufl.algorithms.map_integrands import map_integrand_dags
+from ufl.log import error
+
+
+class ComplexNodeRemoval(MultiFunction):
+    """Replaces complex operator nodes with their children"""
+    def __init__(self):
+        MultiFunction.__init__(self)
+
+    expr = MultiFunction.reuse_if_untouched
+
+    def conj(self, o, a):
+        return a
+
+    def real(self, o, a):
+        return a
+
+    def imag(self, o, a):
+        return a
+
+    def terminal(self, t, *ops):
+        if isinstance(t, ComplexValue):
+            error('Cannot cast ComplexValues to FloatValues!')
+        else:
+            return t
+
+
+def remove_complex_nodes(expr):
+    """Replaces complex operator nodes with their children"""
+    return map_integrand_dags(ComplexNodeRemoval(), expr)
```

Comment:
I'm confused by this, it's not just saying "If I have an expression `R + B*I`, I replace the `B` by zero".  So does it DTRT?

Reply:
No, it's not saying that. It's removing real, imag and conj nodes so that these don't appear in expressions in real mode.

Reply:
Is there a situation in which in "real mode" an `Imag` node is introduced by UFL (rather than the user)?  If the only way we can get `Imag` is if the user typed it, should we raise an error if seeing one here?

Incidentally, again here I found the name misleading.

Reply:
What would you prefer? I think the name makes sense.

Reply:
I couldn't think of an immediately better one :(.

Reply:
e.g. `assume_real`